src/meta_schedule/feature_extractor/per_block_feature.cc

```diff
@@ -314,25 +311,109 @@ class PerBlockFeatureCollector : private StmtVisitor {
         cache_line_bytes_(cache_line_bytes),
         arith_intensity_curve_num_samples_(arith_intensity_curve_num_samples) {}
 
+  void VisitStmt_(const BlockRealizeNode* realize) final {
+    if (!scopes_.empty()) {
+      ordered_blocks_.push_back(realize);
+    }
+    scopes_.push_back(realize);
+    dfs_path_.push_back(realize);
+    StmtVisitor::VisitStmt_(realize);
+    dfs_path_.pop_back();
+    scopes_.pop_back();
+    if (scopes_.empty()) {
+      return;
+    }
+    // Get the ancestor loops from inner to outer, up to the parent scope
+    ForVec loops;
+    for (auto iter = dfs_path_.rbegin(); iter != dfs_path_.rend(); ++iter) {
+      const StmtNode* stmt = *iter;
+      if (stmt->IsInstance<ForNode>()) {
+        loops.push_back(static_cast<const ForNode*>(stmt));
+      }
+    }
+    Feature& feature = block_features_[realize];
+    feature.block_realize = realize;
+    feature.group1 = std::make_unique<group1::Feature>(realize, loop_nest_, is_gpu_);
+    if (!scopes_.empty()) {
+      AddArithOpsToScope(&feature.group1->arith_ops);
+    }
+    // TODO: add groups 2-5
+  }
+
   void VisitStmt_(const ForNode* loop) final {
     int64_t auto_unroll;
+    const int64_t* extent = GetLoopIntExtent(loop);
     ForVec* for_vec = loop_nest_.Push(loop, &auto_unroll);
+    if ((extent && (*extent != 1)) || for_vec) {
+      outer_loop_prod_ *= *extent;
+      dfs_path_.push_back(loop);
+      analyzer_.Bind(loop->loop_var, loop->min);
+    }
     StmtVisitor::VisitStmt_(loop);
     loop_nest_.Pop(loop, for_vec, auto_unroll);
+    if ((extent && (*extent != 1)) || for_vec) {
+      outer_loop_prod_ /= *extent;
+      dfs_path_.pop_back();
+    }
   }
 
-  void VisitStmt_(const BlockRealizeNode* realize) final {
-    // TODO
+  void VisitStmt_(const BufferStoreNode* store) final {
+    ICHECK(!scopes_.empty());
+    group1::Feature::ArithOps arith_ops = group1::Feature::ArithOps(store, loop_nest_.prod);
+    AddArithOpsToScope(&arith_ops);
+  }
+
+  void AddArithOpsToScope(group1::Feature::ArithOps* arith_ops) {
+    const BlockRealizeNode* scope = scopes_.back();
+    // The product of the loops up to the parent
+    int64_t prod_loop_extent = 1;
+    for (auto iter = dfs_path_.rbegin(); iter != dfs_path_.rend(); ++iter) {
+      const tir::StmtNode* stmt = *iter;
+      if (stmt == scope) {
+        break;
+      }
+      ICHECK(stmt->IsInstance<tir::ForNode>());
+      const int64_t* extent = GetLoopIntExtent(static_cast<const tir::ForNode*>(stmt));
+      if (*extent >= 1) {
+        prod_loop_extent *= *extent;
+      }
```

Comment:
No need to check if the extent is `>= 1`. On the other hand, null pointer needs careful checking

```suggestion
      if (const int64_t* extent = GetLoopIntExtent(static_cast<const tir::ForNode*>(stmt))) {
        prod_loop_extent *= *extent;
      }
```

Reply:
Addressed :)